frontend/src/modules/auth/actions.ts

```diff
@@ -0,0 +1,116 @@
+import jwtDecode from "jwt-decode";
+
+import api from "../../utils/api";
+import { createNetworkActionTypes, networkAction } from "../../utils/actions";
+import { selectAuthState } from "../../selectors"
+import { isNullish } from "../../utils/functions"
+
+
+export const LOG_OUT = "AUTH.LOG_OUT";
+
+export const PERFORM_AUTH = createNetworkActionTypes("AUTH.PERFORM");
+export const REFRESH_AUTH_TOKEN = createNetworkActionTypes("AUTH.REFRESH_TOKEN");
+
+export const logOut = () => ({ type: LOG_OUT });
+
+const decodeUserID = tokens => ({ tokens, currentUserID: jwtDecode(tokens.access).user_id })
+
+export const performAuth = (username, password) => async (dispatch, getState) => {
+
+    if (getState().auth.isFetching) return false;
+
+    await dispatch(
+        networkAction(
+            PERFORM_AUTH,
+            api.auth.token({ username, password }),
+            { transform: decodeUserID }));
+
+}
+
+/**
+ * This function checks to see if a user is logged in and has access and refresh tokens.
+ * If so, it checks how long it will be before the access token expires. If it is due to
+ * expire then it will try to refresh the token, as long as the refresh token hasn't
+ * also expired.
+ * 
+ * The function will dispatch the LOG_OUT action if it is unable to refresh the tokens.
+ * 
+ * It returns true if the access token is currently valid and can be used to make
+ * api requests, and false otherwise.
+ * 
+ * @returns boolean
+ */
+export const refreshAuthToken = () => async (dispatch, getState) => {
+    // Refresh the access token if it is within 2 minutes of expiring
+    const TIME_BEFORE_TOKEN_EXPIRES = 2 * 60    
+
+    if (getState().auth.isFetching) {
+        return false
+    } 
+
+    // Check token validity
+    const tokens = getState().auth.tokens;
+
+    if (!tokens.access || !tokens.refresh) {
+        // Missing token, should perform auth instead
+        await dispatch(logOut());
+        return false;
+    }
+
+    try {
+        const access = jwtDecode(tokens.access);
+        const refresh = jwtDecode(tokens.refresh);
+
+        const now = Date.now() / 1000;
+
+        if (access.exp > now + TIME_BEFORE_TOKEN_EXPIRES) {
+            // Access token is still valid for another while, don't refresh yet.
+            return true;
```

Comment:
I would tend to return false here instead ...

Reply:
I'm using the return value of the function to know if the token can be used after the refresh check, so I return true if the access token is still valid, or after successfully refreshing the token. The function used to return either the result of the refresh token api call, or false, but that wasn't used anywhere so I "repurposed" the return value. :)